dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java

```diff
@@ -409,113 +406,60 @@ private RefinementIteration doRefinementIteration(ProverEnvironment prover, WMMS
     // ================================================================================================================
     // Special memory model processing
 
-    // This method cuts off negated relations that are dependencies of some
-    // consistency axiom. It ignores dependencies of flagged axioms, as those get
-    // eagerly encoded and can be completely ignored for Refinement.
-    private static Set<Relation> cutRelationDifferences(Wmm targetWmm, Wmm baselineWmm) {
-        // TODO: Add support to move flagged axioms to the baselineWmm
-        Set<Relation> cutRelations = new HashSet<>();
-        Set<Relation> cutCandidates = new HashSet<>();
-        int cutCounter = 0;
-        targetWmm.getAxioms().stream().filter(ax -> !ax.isFlagged())
-                .forEach(ax -> collectDependencies(ax.getRelation(), cutCandidates));
-        for (Relation rel : cutCandidates) {
-            if (rel.getDefinition() instanceof Difference) {
-                Relation sec = ((Difference) rel.getDefinition()).complement;
-                if (!sec.getDependencies().isEmpty() || sec.getDefinition() instanceof Identity
-                        || sec.getDefinition() instanceof CartesianProduct) {
-                    // NOTE: The check for RelSetIdentity/RelCartesian is needed because they appear
-                    // non-derived in our Wmm but for CAAT they are derived from unary predicates!
-                    logger.info("Found difference {}. Cutting rhs relation {}", rel, sec);
-                    cutRelations.add(sec);
-                    Relation baselineCopy = getCopyOfRelation(sec, baselineWmm);
-                    baselineWmm.addConstraint(new ForceEncodeAxiom(baselineCopy));
-                    // We give the cut relations new aliases in the original and the baseline wmm
-                    // so that we can match them later by name.
-                    targetWmm.addAlias("cut#" + cutCounter, sec);
-                    baselineWmm.addAlias("cut#" + cutCounter, baselineCopy);
-                    cutCounter++;
+    private static RefinementModel generateRefinementModel(Wmm original) {
+        // We cut (i) negated relations (if derived) and (ii) negated axioms
+        final Set<Constraint> constraintsToCut = new HashSet<>();
+        for (Constraint c : original.getConstraints()) {
+            if (c instanceof Axiom ax && ax.isNegated()) {
+                constraintsToCut.add(ax);
+            } else if (c instanceof Difference diff) {
+                final Relation sub = diff.getSubtrahend();
+                final Definition subDef = sub.getDefinition();
+                if (!sub.getDependencies().isEmpty() || subDef instanceof SetIdentity || subDef instanceof CartesianProduct) {
+                    constraintsToCut.add(subDef);
                 }
```

Comment:
The dependency check is not for the diff but the RHS of the diff.

I guess the word "semi-derived" makes sense. The two checks are there because CAAT treats them as derived whereas they are base in our `Wmm` (since we have no unary predicates but CAAT has).
It certainly would make sense to add fences as well, although CAAT currently treats them as a base relation but one over which it cannot compute negations.
I would still prefer a rewrite of the form `fencerel(X) -> po;[X];po` and avoid all special reasoning, but that is orthogonal.

Btw. this code would also cut `data/addr/ctrl` if we decide to do so.